database/sql/V048__create_settlement_rewards.sql

```diff
@@ -0,0 +1,16 @@
+ALTER TABLE settlements
+  ADD COLUMN id bytea default NULL,
+  ADD COLUMN gas_usd integer default NULL,
+  ADD COLUMN gas_price integer default NULL;
```

Comment:
@vkgnosis - do you think it makes sense to stick these into `settlement_observations` table instead? This way the `settlements` table is just for indexing events and is append only (I think we took a similar strategy with `auction_transaction` table).

We can read this as part of the transaction receipt as we index settlements. It could even replace the existing `auction_transaction` table.

Reply:
Whoops that was a typo: was supposed to say `gas_used`

Reply:
Unresolved the thread as there was an open question directed at @vkgnosis

Reply:
Sorry didn't see that.

Reply:
I agree this should not go into the settlements table because that table can get overriden by event updates and it is clearer separation to put it somewhere else. Don't know what the best place is, but not that table.

Reply:
The event. Contains the transaction receipt with the details of gas used and gas price. Why would we even consider not putting it here. In any case I am fine either way since it is also an "observation" of the settlement.

Reply:
We actually need transaction calldata for extracting the observed fee and surplus values for `settlement_observations`. I would propose not `ALTER`-ing the `settlements` table and instead adding these values to the `settlement_observations` (this should make your life easier @bh2smith as you won't need as many `JOIN`s and we don't need to deal with `NULL` values for old data):

```sql
CREATE TABLE settlement_observations (
  -- the appended settlement ID read from the calldata
  id bytea NOT NULL,
  -- the transaction hash and log index to be able to `JOIN` on the `settlement`
  -- table
  tx_hash bytea NOT NULL,
  log_index bigint NOT NULL,
  -- the gas characteristics observed from the transaction receipt
  cumulative_gas_used numeric(78,0) NOT NULL,
  effective_gas_price numeric(78,0) NOT NULL;
  -- the surplus observed from the transaction calldata, and converted to ETH
  -- with the auction external prices.
  surplus numeric(78,0) NOT NULL,
  fee numeric(78,0) NOT NULL
);
```

And, we should also store auction external prices for traded tokens, as they are important to keep around and will be needed for indexing settlement transactions (for computing surplus and fee in ETH, we can also use the `auctions` table, but that gets cleaned up periodically and I'm reluctant to do so for that reason):

```sql
CREATE TABLE auction_prices (
  auction_id bigint NOT NULL,
  token bytea NOT NULL,
  price numeric(78,0) NOT NULL,
);
```

I'm not sure how to optimize storage for `auction_prices`, as we only need them for traded tokens, but given the co-location model we don't have a reliable way for determining which tokens we need to store ATM. (We can always incentivize solvers to be upfront about this by saying "price is 0 for surplus and fee computation for tokens that you did not say were traded" - so they get fewer rewards in that case).

Reply:
I think this thread has been very helpful and brought forth some great points. Have updated the schema (and now there is no need to alter `settlements` table at all). I hope we are getting closer to a reasonable model for this CIP.

Going to close this issue for now. Please feel free to add further comments on the recently proposed schema.

Reply:
> I'm not sure how to optimize storage for auction_prices, as we only need them for traded tokens

The main concern with adding full list of external prices is the storage right? 
On each auction cut we can store full list of external prices to `auction_prices`. After transaction being mined, when we update the `settlement_observations` we can also update the `auction_prices` entry and eliminate not traded tokens.